onnxruntime/core/providers/cpu/reduction/reduction_ops.cc

```diff
@@ -137,7 +137,20 @@ bool PrepareForReduce(OpKernelContext* ctx,
   ORT_ENFORCE(input_tensor_ptr != nullptr);
   const Tensor& input = *input_tensor_ptr;
 
-  size_t ndim = input.Shape().GetDims().size();
+  size_t ndim = input.Shape().NumDimensions();
+
+  // Scalar tensor
+  if (ndim == 0) {
+    auto size = input.Shape().Size();
+    assert(size == 1);
+    transposedInputData.resize(size, 0);
+    T* to_data = &transposedInputData[0];
+    *to_data = *input.Data<T>();
+    block_size = blocks = 1;
+    *reducedTensor = ctx->Output(0, input.Shape());
+    return false;
```

Comment:
This should check check_no_transpose. We don't need a copy and 'transpose' a single element if that is true.

Reply:
I looked into that and there are a lot of cases when the check_no_transpose  is  false, but the calling code is using transposedInputData unconditionally. Am I missing anything? Specifically: ReduceL1, ReduceL2, ReduceLogSum, ReduceLogSumExp, Reduce, ProdReduceSumSquare, ArgMax, ArgMin.

---
In reply to: [394737463](https://github.com/microsoft/onnxruntime/pull/3260#discussion_r394737463) [](ancestors = 394737463)